specs/atp: re-write specs intro (#119)
bluesky-social/atproto-website: content/specs/atp.md

---
title: AT Protocol
summary: Specification for the Authenticated Transfer Protocol (AT Protocol)
---

# AT Protocol

The Authenticated Transfer Protocol (AT Protocol or atproto) is a generic federated protocol for building open social media applications. Some recurring themes and features are:

- self-authenticating data and identity, allowing seamless account migrations and redistribution of content
- design for "big world" use cases, scaling to billions of accounts
- delegated authority over application-layer schemas and aggregation infrastructure
- re-use of existing data models from the dweb protocol family and network primitives from the web platform

## Protocol Structure

**Identity:** account control is rooted in stable [DID](/specs/did) identifiers, which can rapidly resolved to determine the current service provider location and cryptographic keys associated with the account. [Handles](/specs/handle) provide a more human-recognizable and mutable identifier for accounts.

> **Data:** public content is stored in content-addressed and cryptographically verifiable [Repositories](/specs/repository). data records and network messages all conform to a unified [Data Model](/specs/data-model) ([IPLD](https://ipld.io/docs/data-model/), with [CBOR](https://en.wikipedia.org/wiki/CBOR) and JSON representations).

Reviewer: ```suggestion
**Data:** public content is stored in content-addressed and cryptographically verifiable [Repositories](/specs/repository). Data records and network messages all conform to a unified [Data Model](/specs/data-model) ([IPLD](https://ipld.io/docs/data-model/), with [CBOR](https://en.wikipedia.org/wiki/CBOR) and JSON representations).
```

**Network:** HTTP client-server and server-server [APIs](/specs/xrpc) are described with Lexicons, as are WebSocket [Event Streams](/specs/event-stream). Individual records can be referenced across the network by [AT URI](/specs/at-uri-scheme). A Personal Data Server (PDS) acts as an account's trusted agent in the network, routes client network requests, and hosts repositories. A Big Graph Server (BGS) crawls many repositories and outputs a unified event firehose.

> **Application:** APIs and record schemas for applications built on atproto are specified in [Lexicons](/specs/lexicon), which are are referenced by [Namespaced Identifiers](/specs/nsid) (NSIDs). Application-specific aggregations (such as search) are provided by an Application View (AppView) service. Clients can include mobile apps, desktop software, or web interfaces.

Reviewer 2: ```suggestion
**Application:** APIs and record schemas for applications built on atproto are specified in [Lexicons](/specs/lexicon), which are referenced by [Namespaced Identifiers](/specs/nsid) (NSIDs). Application-specific aggregations (such as search) are provided by an Application View (AppView) service. Clients can include mobile apps, desktop software, or web interfaces.
```

Reviewer: ```suggestion
**Application:** APIs and record schemas for applications built on atproto are specified in [Lexicons](/specs/lexicon), which are referenced by [Namespaced Identifiers](/specs/nsid) (NSIDs). Application-specific aggregations (such as search) are provided by an Application View (App View) service. Clients can include mobile apps, desktop software, or web interfaces.
```

> The AT Protocol itself does not specify common social media conventions like "follows" or "avatars", leaving these to application-level Lexicons. The `com.atproto.*` Lexicons provide common APIs for things like account signup and login. These could be considered part of AT Protocol itself, though they can also be extended or replaced over time as needed. Bluesky is a Twitter-style microblogging application built on top of AT Protocol, with lexicons under the `app.bsky.*` namespace.

Reviewer: ```suggestion
The AT Protocol itself does not specify common social media conventions like follows or avatars, leaving these to application-level Lexicons. The `com.atproto.*` Lexicons provide common APIs for things like account signup and login. These could be considered part of AT Protocol itself, though they can also be extended or replaced over time as needed. Bluesky is a microblogging social app built on top of AT Protocol, with lexicons under the `app.bsky.*` namespace.
```

> ## Cryptography

Reviewer 2: kinda feel that this could warrant it's own doc

Reviewer 2: gets a bit in the weeds for the overview docs

Author: @dholms: I felt like it was important to at least mention these at a high-level in the first page, to clarify the scope of these spec docs.

We will definitely need some additional docs for both the com.atproto "built-ins" and the app.bsky application stuff. They are only partially self-documenting via Lexicon.

Author: oh, I see you meant the Cryptography section (github PR review page wasn't clear). Yeah, cryptography could be a separate page.

A common pattern when signing data in atproto is to encode the data in DAG-CBOR, hash the CBOR bytes with SHA-256, yielding raw bytes (not a hex-encoded string), and then sign the hash bytes.

The set of supported cryptographic systems is expected to evolve slowly. There are significant interoperability and implementation advantages to having as few systems as possible at any point in time.

## Protocol Extension and Applications

AT Protocol was designed from the beginning to balance stability and interoperation against flexibility for third-party application development.

The core protocol extension mechanism is development of new Lexicons under independent namespaces. Lexicons can declare new repository record schemas (stored in collections by NSID), new HTTP API endpoints, and new event stream endpoints and message types. It is also expected that new applications might require new network aggregation services ("AppViews") and client apps (eg, mobile apps or web interfaces).

> It is expected that third parties will re-use Lexicons and record data across namespaces. For example, new applications are welcome to build on top of the social graph records specified in the `app.bsky.*` Lexicons, as long as they comply with the schemas controlled by the `bsky.app` authority.

Reviewer: ```suggestion
It is expected that third parties will reuse Lexicons and record data across namespaces. For example, new applications are welcome to build on top of the social graph records specified in the `app.bsky.*` Lexicons, as long as they comply with the schemas controlled by the `bsky.app` authority.
```
